tooling/operation/get-parameters-as-json-schema.js

```diff
@@ -15,173 +15,291 @@ const types = {
+  'deprecated',
+];
+
+function encodePointer(str) {
+  // These characters need to be encoded to be allowed within a JSON pointer.
+  return str.replace('~', '~0').replace('/', '~1');
+}
 
-  const cleanupSchemaDefaults = (originType, obj, prevProp = false, prevProps = []) => {
-    Object.keys(obj).forEach(prop => {
-      // Since this method is recursive, let's reset our states when we're first processing a new property tree.
-      if (!prevProp) {
-        prevProps = [];
+function isPrimitive(val) {
+  return typeof val === 'string' || typeof val === 'number' || typeof val === 'boolean';
+}
+
+function isPolymorphicSchema(schema) {
+  return 'allOf' in schema || 'anyOf' in schema || 'oneOf' in schema;
+}
+
+function isRequestBodySchema(schema) {
+  return 'content' in schema;
+}
+
+function searchForExampleByPointer(pointer, examples = []) {
+  if (!examples.length || !pointer.length) {
+    return undefined;
+  }
+
+  const locSplit = pointer.split('/').filter(Boolean).reverse();
+  const pointers = [];
+
+  let point = '';
+  for (let i = 0; i < locSplit.length; i += 1) {
+    point = `/${locSplit[i]}${point}`;
+    pointers.push(point);
+  }
+
+  let example;
+  const rev = [...examples].reverse();
+
+  // This code has some quirks! If a deeply nested property shares the same name as an example that's further up the
+  // stack (like `tags.id` and an example for `id`), there's a chance that it'll be misidentified as having an example
+  // and receive the wrong example value.
+  //
+  // Any example is usually better than no example though, so while it's quirky behavior it shouldn't raise immediate
+  // cause for alarm.
+  for (let i = 0; i < pointers.length; i += 1) {
+    for (let ii = 0; ii < rev.length; ii += 1) {
+      let schema = rev[ii];
+      if ('example' in schema) {
+        schema = schema.example;
+      } else {
+        schema = schema.examples[Object.keys(schema.examples).shift()].value;
+      }
+
+      example = jsonpointer.get(schema, pointers[i]);
+      if (example !== undefined) {
+        break;
       }
+    }
 
-      if (obj[prop] === null) {
-        // If the item is null, just carry on. Why do this in addition to `typeof obj[prop] == object`? Because
-        // `typeof null` equates to `object` for "legacy reasons" apparently.
-        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/null
-      } else if (typeof obj[prop] === 'object' && !Array.isArray(obj[prop])) {
-        // If we have a `properties` object, but no adjacent `type`, we know it's an object so just cast it as one.
-        if (prop === 'properties' && !('type' in obj)) {
-          if (prevProp && prevProp === 'properties') {
-            // Only add a type if the previous prop isn't also named `properties`!
+    if (example !== undefined) {
+      break;
+    }
+  }
+
+  return example;
+}
+
+function constructSchema(data, prevSchemas = [], currentLocation = '') {
```

Comment:
Can you add some comments to this function describing what it does, and why we want to do it? This file is pretty dense and I think more comments will help clear it up.